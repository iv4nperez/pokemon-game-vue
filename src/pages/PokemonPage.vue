<template>
    <h1 v-if="!pokemon">Espere por favor...</h1>

    <div v-else>
        <h1>Quien es este pokemon?</h1>

        <PokemonPicture
            :pokemonId="pokemon.id"
            :showPokemon="showPokemon"
        />

        <PokemonOptios
            :pokemons="pokemonArr"
            @selection="checkAnser"
        />

       <template v-if="showAnswer" class="fade-in">
            <h2>{{ message }}</h2>
            <button @click="newGame">
                Nuevo Juego
            </button>
       </template>


    </div>
</template>
<script>
import { defineAsyncComponent } from '@vue/runtime-core'
import getPokemonOptions from '@/helpers/getPokemonOptions'



export default {
    components: {
        PokemonPicture  : defineAsyncComponent( () => import('@/components/PokemonPicture.vue') ),
        PokemonOptios  : defineAsyncComponent( () => import('@/components/PokemonOptions.vue') )
    },
    data(){
        return{
            pokemonArr: [],
            pokemon: null,
            showPokemon: false,
            showAnswer: false,
            message:''
        }
    },
    methods: {
        async mixPokemonArray () {
            this.pokemonArr = await getPokemonOptions()

            const rndInt = Math.floor( Math.random() * 4 )
            this.pokemon = this.pokemonArr[ rndInt ]
        },
        checkAnser ( selectedId ) {
            this.showPokemon = true
            this.showAnswer = true

            if ( selectedId === this.pokemon.id) {
                this.message = `Correcto ${ this.pokemon.name }`
            } else {
                this.message = `Oops, era ${ this.pokemon.name }`
            }
        },
        newGame () {
            this.showPokemon = false
            this.showAnswer = false
            this.message = ''
            this.pokemonArr = []
            this.pokemon = null

            this.mixPokemonArray()
        }
    },
    mounted () {
        this.mixPokemonArray();
    }
}
</script>