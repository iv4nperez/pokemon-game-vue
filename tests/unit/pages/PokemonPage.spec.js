import { shallowMount, mount } from "@vue/test-utils";
import PokemonPage from "@/pages/PokemonPage";
import { pokemons } from "../mocks/pokemons.mock";



describe('PokemonPage Component', () => {
    
    let wrapper

    beforeEach(() => {
        wrapper =  shallowMount( PokemonPage )
    })

    
    // test('debde de hacer match con el snapshot', () => {
        
    // })
    
    test('debe de llamar el mixPokemonArray al montar',  () => {
        const mixPokemonArraySpy = jest.spyOn( PokemonPage.methods, 'mixPokemonArray' )
        shallowMount( PokemonPage )

        expect( mixPokemonArraySpy ).toHaveBeenCalled()
    })
    
    test('debde de hacer match con el snapshotcuando carga el componente', () => {
        
        const wrapper =  shallowMount( PokemonPage, {
            data(){
                return{
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message:'',
                    diabledOptions: false
                }
            }
        })


        expect( wrapper.html() ).toMatchSnapshot()
    })

    test('debe de mostrar los componentes de pokemon picture y pokemon options', () => {
        
        
        const wrapper =  shallowMount( PokemonPage, {
            data(){
                return{
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message:'',
                    diabledOptions: false
                }
            }
        })

        const options = wrapper.find('pokemon-optios-stub')
        const picture = wrapper.find('pokemon-picture-stub')


        expect( options.exists() ).toBeTruthy();
        expect( picture.exists() ).toBeTruthy();

        expect(  options.attributes('pokemons') ).toBeTruthy()
        expect(  picture.attributes('pokemonid') ).toBe( '1' )


    })
    
    test('Pruebas del componente CheckAnswer', async () => {
        
        const wrapper =  shallowMount( PokemonPage, {
            data(){
                return{
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message:'',
                    diabledOptions: false
                }
            }
        })

        await wrapper.vm.checkAnser( 1 )

        expect( wrapper.find('h2').exists() ).toBeTruthy()
        expect( wrapper.vm.showPokemon ).toBeTruthy()
        expect( wrapper.find('h2').text() ).toBe( `Correcto ${ pokemons[0].name }` )

        await wrapper.vm.checkAnser( 2 )

        expect( wrapper.vm.message ).toBe( `Oops, era ${ pokemons[0].name }` )
    })
    
})
