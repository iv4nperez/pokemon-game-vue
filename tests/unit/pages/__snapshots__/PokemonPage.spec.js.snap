// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Component debde de hacer match con el snapshotcuando carga el componente 1`] = `
<div>
  <h1>Quien es este pokemon?</h1>
  <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub>
  <pokemon-optios-stub pokemons="[object Object],[object Object],[object Object],[object Object]" diabledoptions="false"></pokemon-optios-stub>
  <!--v-if-->
</div>
`;
